components/gfx/platform/macos/font_template.rs

```diff
@@ -2,182 +2,69 @@
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt.debug_struct("FontTemplateData")
-            .field("ctfont", &self.ctfont)
-            .field("identifier", &self.identifier)
-            .field(
-                "font_data",
-                &self
-                    .font_data
-                    .read()
-                    .unwrap()
-                    .as_ref()
-                    .map(|bytes| format!("[{} bytes]", bytes.len())),
-            )
-            .finish()
-    }
-}
+/// A [`HashMap`] of cached [`CTFont`] for a single [`FontIdentifier`]. There is one [`CTFont`]
+/// for each cached font size.
+type CachedCTFont = HashMap<Au, CTFont>;
 
-unsafe impl Send for FontTemplateData {}
-unsafe impl Sync for FontTemplateData {}
-
-impl FontTemplateData {
-    pub fn new(
-        identifier: FontIdentifier,
-        font_data: Option<Vec<u8>>,
-    ) -> Result<FontTemplateData, IoError> {
-        Ok(FontTemplateData {
-            ctfont: CachedCTFont(Mutex::new(HashMap::new())),
-            identifier,
-            font_data: RwLock::new(font_data.map(Arc::new)),
-        })
-    }
-
-    /// Retrieves the Core Text font instance, instantiating it if necessary.
-    pub fn ctfont(&self, pt_size: f64) -> Option<CTFont> {
-        let mut ctfonts = self.ctfont.lock().unwrap();
+pub(crate) trait CoreTextFontTemplateMethods {
+    /// Retrieves a [`CTFont`] instance, instantiating it if necessary if it is not
+    /// stored in the shared Core Text font cache.
+    fn core_text_font(&self, pt_size: f64) -> Option<CTFont>;
+}
 
-        let entry = ctfonts.entry(Au::from_f64_px(pt_size));
-        match entry {
-            Entry::Occupied(entry) => return Some(entry.get().clone()),
-            Entry::Vacant(_) => {},
+impl CoreTextFontTemplateMethods for FontTemplateRef {
+    fn core_text_font(&self, pt_size: f64) -> Option<CTFont> {
+        //// If you pass a zero font size to one of the Core Text APIs, it'll replace it with
+        //// 12.0. We don't want that! (Issue #10492.)
+        let clamped_pt_size = pt_size.max(0.01);
+        let au_size = Au::from_f64_px(clamped_pt_size);
+
+        let cache = CTFONT_CACHE.get_or_init(Default::default);
+        let identifier = self.borrow().identifier.clone();
+        {
+            let cache = cache.read();
+            if let Some(core_text_font) = cache
+                .get(&identifier)
+                .and_then(|identifier_cache| identifier_cache.get(&au_size))
+            {
+                return Some(core_text_font.clone());
+            }
         }
 
-        // If you pass a zero font size to one of the Core Text APIs, it'll replace it with
-        // 12.0. We don't want that! (Issue #10492.)
-        let clamped_pt_size = pt_size.max(0.01);
+        let mut cache = cache.write();
+        let identifier_cache = cache.entry(identifier).or_insert_with(Default::default);
+        if let Some(core_text_font) = identifier_cache.get(&au_size) {
```

Comment:
Do we perform this check again here because the `identifier_cache` might been updated after the read lock is released and thus it is not guaranteed to be a "cache miss" when we acquire the write lock? If so, do you think it would be helpful to document this scenario in a comment?